ScarletEngine/Sources/AchillesPhysics/Private/ColliderSystem.cpp

```diff
@@ -0,0 +1,117 @@
+
+	void SphereVsSphereColliderSystem::FixedUpdate() const
+	{
+		ZoneScoped
+		// Move collider
+		for (const auto& [Entity, Rb, Sphere] : GetEntities<RigidBodyComponent, SphereColliderComponent>())
+		{
+			Sphere->Pos += Rb->Velocity * static_cast<float>(FIXED_UPDATE_S);
+		}
+
+		// Compute intersection
+		const auto& Entities = GetEntities<Transform, SphereColliderComponent, RigidBodyComponent>();
+		const size_t Size = Entities.size();
+
+		if (Size == 0)
+		{
+			return;
+		}
+
+		for (size_t i = 0; i < Size - 1; i++)
+		{
+			const auto& [EntityA, TransA, SphereA, RbA] = Entities[i];
+			for (size_t j = i + 1; j < Size; j++)
+			{
+				const auto& [EntityB, TransB, SphereB, RbB] = Entities[j];
+				const IntersectionData IntersectionDepth = GetIntersection(SphereA, SphereB);
+
+				if (IntersectionDepth.Distance < 0.f)
+				{
+					const glm::vec3 Fi = - (glm::length(RbA->Velocity - RbB->Velocity) * IntersectionDepth.Direction) / (static_cast<float>(FIXED_UPDATE_S) * (1 / RbA->Mass + 1 / RbB->Mass));
+
+					SolveIntersection(RbA, TransA, SphereA, SphereB, Fi, IntersectionDepth.Direction * IntersectionDepth.Distance);
+					SolveIntersection(RbB, TransB, SphereB, SphereA, -Fi, -IntersectionDepth.Direction * IntersectionDepth.Distance);
+				}
+			}
+		}
+	}
+
+	const IntersectionData SphereVsSphereColliderSystem::GetIntersection(const SphereColliderComponent* SphereA, const SphereColliderComponent* SphereB) const
+	{
+		const glm::vec3 CentDist = SphereB->Pos - SphereA->Pos;
+		const float RadDist = SphereB->Radius + SphereA->Radius;
+		IntersectionData Data;
+		Data.Direction = glm::normalize(CentDist);
+		Data.Distance = glm::length(CentDist) - RadDist;
+		return Data;
+	}
+
+	void SphereVsSphereColliderSystem::SolveIntersection
+		(RigidBodyComponent* Rb, Transform* Trans, SphereColliderComponent* Sphere, SphereColliderComponent* OtherSphere, const glm::vec3 Fi, const glm::vec3 newPos) const
+	{
+		Rb->Force += Fi;
+		Rb->Velocity *= (1.f - OtherSphere->FrictionCoefficient);
+		Trans->Position += newPos;
+		Sphere->Pos = Trans->Position;
+		if (Rb->UsesGravity)
+		{
+			Rb->Force -= Rb->Gravity * Rb->Mass;
+		}
+	}
+
+	void PlaneVsSphereColliderSystem::FixedUpdate() const
+	{
+		ZoneScoped
+		// Get Intersection
+		for (const auto& [EntP, Plane] : GetEntities<PlaneColliderComponent>())
+		{
+			for (const auto& [EntS, TransSphere, Sphere, RbSphere] : GetEntities<Transform, SphereColliderComponent, RigidBodyComponent>())
+			{
+				const IntersectionData IntersectionDepth = GetIntersection(Plane, Sphere);
+
+				if (IntersectionDepth.Distance < 0.f && RbSphere)
+				{
+					const glm::vec3 NewPos = Plane->Normal * -IntersectionDepth.Direction;
+
+					TransSphere->Position += NewPos;
+					RbSphere->Velocity = (1.f - Plane->FrictionCoefficient) * glm::reflect(RbSphere->Velocity, Plane->Normal);
+
+					Sphere->Pos += NewPos;
+
+					if (RbSphere->UsesGravity)
+					{
+						RbSphere->Force -= RbSphere->Gravity * RbSphere->Mass;
+					}
```

Comment:
Instead of using a boolean to determine if this uses gravity, would it not be better to just set the gravity to the zero vector and do the calculation anyways?